capsules/extra/src/app_loader.rs

```diff
@@ -0,0 +1,470 @@
+                app.pending_command = false;
+                // Signal the app.
+                kernel_data
+                    .schedule_upcall(upcall::SETUP_DONE, (into_statuscode(result), 0, 0))
+                    .ok();
+            });
+        });
+    }
+
+    /// Let the app know we are done writing the block of data
+    fn write_done(&self, result: Result<(), ErrorCode>, buffer: &'static mut [u8], length: usize) {
+        // Switch on which user of this capsule generated this callback.
+        self.current_process.map(|processid| {
+            let _ = self.apps.enter(processid, move |app, kernel_data| {
+                // Replace the buffer we used to do this write.
+                self.buffer.replace(buffer);
+                app.pending_command = false;
+
+                // And then signal the app.
+                kernel_data
+                    .schedule_upcall(upcall::WRITE_DONE, (into_statuscode(result), length, 0))
+                    .ok();
+            });
+        });
+    }
+
+    fn finalize_done(&self, result: Result<(), ErrorCode>) {
+        self.current_process.map(|processid| {
+            let _ = self.apps.enter(processid, move |app, kernel_data| {
+                // And then signal the app.
+                app.pending_command = false;
+
+                self.current_process.take();
+                kernel_data
+                    .schedule_upcall(upcall::FINALIZE_DONE, (into_statuscode(result), 0, 0))
+                    .ok();
+            });
+        });
+    }
+
+    fn abort_done(&self, result: Result<(), ErrorCode>) {
+        self.current_process.map(|processid| {
+            let _ = self.apps.enter(processid, move |app, kernel_data| {
+                // And then signal the app.
+                app.pending_command = false;
+
+                self.current_process.take();
+                kernel_data
+                    .schedule_upcall(upcall::ABORT_DONE, (into_statuscode(result), 0, 0))
+                    .ok();
+            });
+        });
+    }
+}
+
+impl kernel::dynamic_binary_storage::DynamicProcessLoadClient for AppLoader<'_> {
+    /// Let the requesting app know we are done loading the new process
+    fn load_done(&self, result: Result<(), ProcessLoadError>) {
+        let status_code = match result {
+            Ok(()) => 0,
+            Err(e) => match e {
+                ProcessLoadError::NotEnoughMemory => 9,
+                ProcessLoadError::MpuInvalidFlashLength => 6,
+                ProcessLoadError::MpuConfigurationError => 1,
+                ProcessLoadError::MemoryAddressMismatch { .. } => 1,
+                ProcessLoadError::NoProcessSlot => 1,
+                ProcessLoadError::BinaryError(_) => 1,
+                ProcessLoadError::CheckError(_) => 1,
+                // This error is usually a result of bug in the kernel
+                // so we return Powered OFF error, because that is unlikely.
+                ProcessLoadError::InternalError => 4,
```

Comment:
Please use named `ErrorCode` variants and the `into_statuscode` function here. It'd be great if you could document this mapping in the interface docs! (Like on the `command` function)

https://docs.tockos.org/kernel/errorcode/fn.into_statuscode.html